### ISSUE TITLE
Add support for derangement streams

### ISSUE THREAD

**Issue by siordache** [beryx/streamplify#2]
This is an _up-for-grabs_ issue, so hurry up and grab it!

Your task is to add support for [derangements](https://en.wikipedia.org/wiki/Derangement) streams.

A derangement stream can be easily obtained by filtering out the permutations that have items appearing on their original positions:

```
new Permutations(n)
        .stream()
        .filter(perm -> Arrays.stream(perm).noneMatch(i -> perm[i] == i))
```

This is, however, a computationally expensive solution. Therefore, your job is to implement efficient classes for derangement streams.

Consult the implementation of the other combinatorics classes (permutations, combinations, Cartesian product) and structure your code in a similar way:
- put your classes in a package named `org.beryx.streamplify.derangement`.
- create a subclass of [LongIndexedSpliterator](../blob/master/streamplify/src/main/java/org/beryx/streamplify/LongIndexedSpliterator.java)  called `LongDerangements`, which can be used for derangements with a cardinality that fits in a long (that is, for sets with at most 21 elements).
  _Hint_: use [LongPermutations.java](../blob/master/streamplify/src/main/java/org/beryx/streamplify/permutation/LongPermutations.java) as template for this class.
- create a subclass of [BigIntegerIndexedSpliterator](../blob/master/streamplify/src/main/java/org/beryx/streamplify/BigIntegerIndexedSpliterator.java) called `BigIntegerDerangements`, which allows derangements of virtually any cardinality.
  _Hint_: use [BigIntegerPermutations.java](../blob/master/streamplify/src/main/java/org/beryx/streamplify/permutation/BigIntegerPermutations.java) as template for this class.
- create a [StreamableProxy](../blob/master/streamplify/src/main/java/org/beryx/streamplify/StreamableProxy.java)  called`Derangements`, which delegates to either _LongDerangements_ or _BigIntegerDerangements_.
  _Hint_: use [Permutations.java](../blob/master/streamplify/src/main/java/org/beryx/streamplify/permutation/Permutations.java) as template for this class.
- create an abstract class called `DerangementSupplier`, which implements [IntArraySupplier](../blob/master/streamplify/src/main/java/org/beryx/streamplify/IntArraySupplier.java) and has two concrete subclasses: the static nested classes `Long` and `BigInteger`.
  _Hint_: use [PermutationSupplier.java](../blob/master/streamplify/src/main/java/org/beryx/streamplify/permutation/PermutationSupplier.java) as template for this class.
  (This is the most challenging class to be implemented, because you have to come up with efficient algorithms for the methods _computeNext()_ and _unrank()_.)

Your pull request should also contain:
- [Spock](http://spockframework.org/) unit tests - see for example [PermutationsSpec.groovy](../blob/master/streamplify/src/test/groovy/org/beryx/streamplify/PermutationsSpec.groovy)
- examples - should be placed in [streamplify-examples](../blob/master/streamplify-examples/src/main/java/org/beryx/streamplify/example)
- documentation (using [Asciidoctor](http://asciidoctor.org/))

In order to prevent that two people work on the same thing, leave a comment here to claim this issue before starting to work on it.

Happy coding, and don't hesitate to ask for clarification and guidance.

**Comment by shahprincea** [comment]
I will give a try.....

**Comment by siordache** [comment]
Grabbed by @shahprincea.

**Comment by shahprincea** [comment]
Dropping out sorry

**Comment by siordache** [comment]
The issue is _up-for-grabs_ again.

**Comment by polmauri** [comment]
I will work on this.

**Comment by siordache** [comment]
Grabbed by @polmauri .

**Comment by siordache** [comment]
Implemented in #4 by @polmauri .

Will be released in 1.1.0.
